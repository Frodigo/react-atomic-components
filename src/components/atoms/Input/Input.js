import React from 'react';
import PropTypes from 'prop-types';

import classes from './Input.module.css';

const Input = (props) => {
    const {
        errorMessage,
        type,
        placeholder,
        state,
        value,
        onChange,
        isTextarea,
        register,
        ...rest
    } = props;

    let inputClass = null;

    switch (state) {
        case 'default': {
            inputClass = classes.input;

            break;
        }

        case 'focus': {
            inputClass = `${classes.input} ${classes.inputFocus}`;

            break;
        }

        case 'error': {
            inputClass = `${classes.input} ${classes.inputError}`;

            break;
        }

        case 'disabled': {
            inputClass = `${classes.input} ${classes.inputDisabled}`;

            break;
        }

        default: {
            inputClass = classes.input;
        }
    }

    const isDisabled = state === 'disabled';
    const shouldDisplayError = errorMessage ? (
        <span className={classes.errorMessage}>{errorMessage.message}</span>
    ) : null;
    const errorClass = errorMessage ? classes.hasError : null;

    if (isTextarea) {
        return (
            <div className={classes.field}>
                <textarea
                    className={`${inputClass} ${classes.textarea} ${errorClass}`}
                    placeholder={placeholder}
                    value={value}
                    onChange={onChange}
                    disabled={isDisabled}
                    ref={register}
                    {...rest}
                    data-testid="input"
                />
                {shouldDisplayError}
            </div>
        );
    }

    return (
        <div className={classes.field}>
            <input
                className={`${inputClass} ${errorClass}`}
                type={type}
                placeholder={placeholder}
                value={value}
                onChange={onChange}
                disabled={isDisabled}
                ref={register}
                {...rest}
                data-testid="input"
            />
            {shouldDisplayError}
        </div>
    );
};

export default Input;

Input.propTypes = {
    errorMessage: PropTypes.object,
    isTextarea: PropTypes.bool,
    onChange: PropTypes.func,
    placeholder: PropTypes.string,
    register: PropTypes.func,
    state: PropTypes.oneOf(['default', 'focus', 'error', 'disabled']),
    type: PropTypes.oneOf([
        'text',
        'email',
        'number',
        'password',
        'search',
        'tel'
    ]).isRequired,
    value: PropTypes.string
};

Input.defaultProps = {
    type: 'text',
    state: 'default',
    isTextarea: false
};
