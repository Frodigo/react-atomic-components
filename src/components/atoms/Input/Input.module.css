.field {
    margin-bottom: var(--spacing-md);
}

.input {
    background-color: var(--color-white);
    border: solid 2px var(--color-alternate);
    border-radius: 0;
    color: var(--color-tiertary);
    font-size: 16px;
    height: 50px;
    outline: none;
    padding: 0 var(--spacing-md);
    transition: .3s ease-in-out border;
    width: 100%;
}

.hasError {
    border-color: var(--color-error);
}

.textarea {
    height: auto;
    min-height: 200px;
    padding: var(--spacing-md);
    resize: none;
}

.input::-webkit-input-placeholder,
.input:-ms-input-placeholder,
.input::placeholder {
    color: var(--color-gray-light);
    opacity: 1;
}


.inputFocus,
.input:focus {
    border-color: var(--color-secondary);
}

.inputError {
    border-color: var(--color-red);
    color: var(--color-red);
}

.inputDisabled {
    border-color: var(--color-gray-light);
    color: var(--color-gray-light);
    pointer-events: none;
}


.errorMessage {
    color: var(--color-error);
    display: block;
    margin-top: 5px;
    padding-left: var(--spacing-xs);
    text-align: left;

}