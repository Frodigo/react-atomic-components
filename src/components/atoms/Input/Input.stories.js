import React from 'react';

import { default as InputComponent } from './Input';

export default {
    title: 'Atoms/Form/Inputs',
    component: InputComponent
};
const Template = (args) => <InputComponent {...args} />;

export const DefaultInput = Template.bind({});
DefaultInput.args = {
    placeholder: 'Type here...'
};

export const FocusedInput = Template.bind({});
FocusedInput.args = {
    value: 'John',
    state: 'focus'
};

export const ErroredInput = Template.bind({});
ErroredInput.args = {
    value: 'John',
    state: 'error',
    errorMessage: 'This is a required field.'
};

export const Textarea = Template.bind({});
Textarea.args = {
    value: 'Lorem ipsum dolor sit amet',
    isTextarea: true
};
