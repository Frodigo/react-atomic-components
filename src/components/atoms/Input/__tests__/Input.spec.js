import React from 'react';
import { render, getByTestId, getByText } from '@testing-library/react';

import Input from '../Input';

describe('Input should be rendered', () => {
    test('Input renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            onChange: jest.fn()
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(
            getByTestId(container, 'input').getAttribute('type')
        ).toStrictEqual('text');
        expect(
            getByTestId(container, 'input').getAttribute('placeholder')
        ).toStrictEqual('Type here...');
    });

    test('Input error renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            type: 'email',
            onChange: jest.fn(),
            state: 'error'
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(
            getByTestId(container, 'input').classList.contains('inputError')
        ).toBe(true);
    });

    test('Input focus renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            type: 'email',
            onChange: jest.fn(),
            state: 'focus'
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(
            getByTestId(container, 'input').classList.contains('inputFocus')
        ).toBe(true);
    });

    test('Input disabled renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            type: 'email',
            onChange: jest.fn(),
            state: 'disabled'
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(
            getByTestId(container, 'input').classList.contains('inputDisabled')
        ).toBe(true);
        expect(
            getByTestId(container, 'input').getAttribute('disabled')
        ).toBeDefined();
    });

    test('textarea renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            isTextarea: true,
            onChange: jest.fn()
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(
            getByTestId(container, 'input').classList.contains('textarea')
        ).toBe(true);
    });

    test('error message renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            isTextarea: true,
            errorMessage: {
                message: 'error'
            },
            onChange: jest.fn()
        };
        const { container } = render(<Input {...props} />);

        expect(container).toMatchSnapshot();
        expect(getByText(container, 'error')).toBeDefined();
    });

    test('error class renders correctly', () => {
        const props = {
            placeholder: 'Type here...',
            isTextarea: false,
            errorMessage: {
                message: 'error'
            },
            onChange: jest.fn()
        };
        const { container } = render(<Input {...props} />);

        expect(
            getByTestId(container, 'input').classList.contains('hasError')
        ).toBe(true);
    });
});
