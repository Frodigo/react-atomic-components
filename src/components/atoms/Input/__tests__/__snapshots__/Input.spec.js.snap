// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should be rendered Input disabled renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <input
      class="input inputDisabled null"
      data-testid="input"
      disabled=""
      placeholder="Type here..."
      type="email"
      value=""
    />
  </div>
</div>
`;

exports[`Input should be rendered Input error renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <input
      class="input inputError null"
      data-testid="input"
      placeholder="Type here..."
      type="email"
      value=""
    />
  </div>
</div>
`;

exports[`Input should be rendered Input focus renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <input
      class="input inputFocus null"
      data-testid="input"
      placeholder="Type here..."
      type="email"
      value=""
    />
  </div>
</div>
`;

exports[`Input should be rendered Input renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <input
      class="input null"
      data-testid="input"
      placeholder="Type here..."
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Input should be rendered error message renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <textarea
      class="input textarea hasError"
      data-testid="input"
      placeholder="Type here..."
    />
    <span
      class="errorMessage"
    >
      error
    </span>
  </div>
</div>
`;

exports[`Input should be rendered textarea renders correctly 1`] = `
<div>
  <div
    class="field"
  >
    <textarea
      class="input textarea null"
      data-testid="input"
      placeholder="Type here..."
    />
  </div>
</div>
`;
